virtualization-and-containers/material/exercise_vm_containers_text.md

```diff
@@ -0,0 +1,175 @@
+- You can find more information on the installation procedure in the [Ubuntu Server documentation](https://ubuntu.com/server/docs).
+- [Ubuntu Server download page](https://ubuntu.com/download/server)
+- [VirtualBox Manual](https://www.virtualbox.org/manual/UserManual.html)
+- If you want to set up SSH forwarding to your VM, you can find instructions, for example, [on the codebot homepage](https://codebots.com/docs/ubuntu-18-04-virtual-machine-setup). The instructions are for Ubuntu 18.04, but it also works for Ubuntu 20.04. Note that most of the steps are preconfigured in current VirtualBox releases, but it does not hurt to check whether the settings are set properly.
+
+
+## Virtual Machines Using Vagrant
+
+In the previous section we have set up a VM manually. This was quite tedious. Therefore, we want to automatize this process now by using [Vagrant](https://www.vagrantup.com/) and setting up our own Vagrant box. We will provision a box based on VirtualBox by using shell scripts and the included provisioning tools of Vagrant.
+
+### Tasks (Vagrant)
+
+- Fork the repository [TODO: "Exercise Virtual Machines"](https://gitlab-sim.informatik.uni-stuttgart.de/sse-test-group/exercise-virtual-machines) from GitLab and create a branch to work on the fork. It will contain a `README.md` and a file called `bootstrap.sh`.
+- We want to start from scratch so initialize a new box using `vagrant init` in inside this repository and add the resulting `Vagrantfile` to Git. Then adapt your `Vagrantfile` to incorporate the following settings:
+  - Your virtual machine must be based on the [`ubuntu/focal64` image](https://www.vagrantup.com/docs/boxes#official-boxes).
+  - The name of your VM must be `USERNAME-ubuntu-server`.
+  - The [box version](https://www.vagrantup.com/docs/boxes/versioning) must be `0.1.0`.
+  - The VM must request [384 MB of main memory](https://www.vagrantup.com/docs/providers/virtualbox/configuration).
+  - Run your box with `vagrant up` and make sure that everything works out as expected (`vagrant ssh`). If everything is fine, you can leave the VM and stop it.
+- In the next step we want to provision your virtual machines. That means we want to install additional software, add extra files, and set some additional variables. We want to provision (`config.vm.provision`) the box in several steps.
+  - Create a new file with the name `testfile` and add it to the repository. The file must contain your GitLab username as text. Add the file to your box using the [file provisioner](https://www.vagrantup.com/docs/provisioning/file) and place it in the home directory of the `vagrant` user.
+  - We want an [additional shared folder](https://www.vagrantup.com/docs/synced-folders/basic_usage), besides `/vagrant`, in our virtual machine. The directory where the `Vagrantfile` resides, i.e. `.`, should be mounted as `/mnt/shared/` in your container.
```

Comment:
This is technically not part of `config.vm.provision`. It does not get updated when using `vagrant provision`. 
Maybe we could do two steps here? First pure provision (everything but the shared folder) and they should use `vagrant provision`. And afterwards the shared folder + `vagrant destroy` + `vargant up`?

Reply:
Splitting it into two parts sounds good to me. I will check this in more detail. I must admit I did not use `vagrant provision` when building my machine but rather `vagrant up --provision` or so. My memory is fading. :smile:

Reply:
If you do a plain `vagrant up` when the machine is already running, Vagrant suggests to either use `vagrant provision` or `vagrant up --provision`.

Reply:
I split it up now to first set up the baseline of the VM and then add the environment variables and software using `vagrant provision`/`vagrant up --provision`. This way the workflow is closer to what Vagrant's documentation suggests (set up baseline and the incrementally extend the provisioning).